src/main/java/com/salesforce/cdp/queryservice/util/TokenHelper.java

```diff
@@ -276,4 +337,70 @@ private static Response un_pw_login(
             fillArray(body, (byte)0);
         }
     }
+
+    private static Response key_pair_auth_login(
+            String grantType, String clientId,
+            String userName, String privateKey, String audience, String tokenUrl, OkHttpClient client
+    ) throws TokenException {
+        byte[] grantTypeSegment = (Constants.GRANT_TYPE_NAME + Constants.TOKEN_ASSIGNMENT + grantType)
+                .getBytes(StandardCharsets.UTF_8);
+        byte[] jwsSegment = (Constants.ASSERTION + Constants.TOKEN_ASSIGNMENT +
+                createJwt(clientId, userName, privateKey, audience)).getBytes(StandardCharsets.UTF_8);
+        byte[] separator = Constants.TOKEN_SEPARATOR.getBytes(StandardCharsets.UTF_8);
+
+        byte[] body = Bytes.concat(
+                grantTypeSegment, separator, jwsSegment
+        );
+        try {
+            RequestBody requestBody = RequestBody.create(body, MediaType.parse(Constants.URL_ENCODED_CONTENT));
+            Map<String, String> headers = Collections.singletonMap(Constants.CONTENT_TYPE, Constants.URL_ENCODED_CONTENT);
+            Request request = HttpHelper.buildRequest(Constants.POST, tokenUrl, requestBody, headers);
+            Response response = client.newCall(request).execute();
+            if (!response.isSuccessful()) {
+                log.error("login with user credentials failed with status code {}", response.code());
+                HttpHelper.handleErrorResponse(response, Constants.ERROR_DESCRIPTION);
+            }
+            return response;
+        } catch (IOException e) {
+            log.error("login with user credentials failed", e);
+            throw new TokenException(FAILED_LOGIN, e);
+        } finally {
+            fillArray(grantTypeSegment, (byte)0);
+            fillArray(jwsSegment, (byte)0);
+            fillArray(separator, (byte)0);
+            fillArray(body, (byte)0);
+        }
+    }
+
+    private static String createJwt(String clientId, String userName, String privateKey, String audience) throws TokenException {
+        Instant now = Instant.now();
+        String jwtToken = null;
+        try {
+            RSAPrivateKey rsaPrivateKey = getPrivateKey(privateKey);
+            jwtToken = Jwts.builder()
+                    .setIssuer(clientId)
+                    .setSubject(userName)
+                    .setAudience(audience)
+                    .setIssuedAt(Date.from(now))
+                    .setExpiration(Date.from(now.plus(5l, ChronoUnit.MINUTES)))
```

Comment:
Do we need 5 mins of expiry time for this jwt? Should we lower this

Reply:
Keeping it 2 mins as per: https://help.salesforce.com/s/articleView?id=sf.remoteaccess_oauth_jwt_flow.htm&type=5 

<img width="979" alt="image" src="https://user-images.githubusercontent.com/101810633/160253968-4612f4c4-e4c1-4212-b05e-3a9c7d6cde8c.png">